### ISSUE TITLE
FNX3-16897 ⁃ Pull to Refresh Web Page

### ISSUE THREAD

**Issue by mheubusch** [mozilla-mobile/fenix#3262]
## User Story
As a user, I want to pull to refresh content on any webpage

## Dependencies
* [ ] GV pull  to refresh capability
* [ ] Behind a Nightly feature flag
* [ ] Notify QA for testing, because they requested extra testing on this feature
* [ ]  UI review with  Amy

### Acceptance Criteria
* I can pull to  refresh
* I can see that the design of the pull to refresh UI is aligned with the Fenix design system
* Behind a Feature Flag so QA can spend extra time testing

**Comment by sblatz** [comment]
Isn't this something that @NotWoods already has working on Sample Browser in a-c? 🤔

**Comment by NotWoods** [comment]
Yeah, this is already being tested in Sample Browser & Reference Browser, and in Fenix behind a flag.

**Comment by arps18** [comment]
Is the refresh option available or it is not working in the specific sites?

**Comment by robsmith11** [comment]
Please consider having this gesture be optional. I very rarely have a need to refresh a page and find it easy to accidentally trigger refreshes in other apps with pull-to-refresh when scrolling.

**Comment by cadeyrn** [comment]
> Is the refresh option available or it is not working in the specific sites?

This issue is still open and the comment above your comment says that it's still behind a flag. => No, it's not available in regular builds. You have to set the build flag and compile your own build to get this feature now.

**Comment by kolAflash** [comment]
Maybe related?
- https://github.com/mozilla-mobile/android-components/issues/1824 _Meta: Pull To Refresh_

**Comment by NotWoods** [comment]
Blocked by mozilla-mobile/android-components#3182

**Comment by mheubusch** [comment]
@AmyYLee Just flagging you here as I think this is coming soon and want to make sure you have weighed in on the things you can control within the Material UI component for swipetorefresh, like color and gesture distance. also @lime124 for visibility. Removing myself.

**Comment by lime124** [comment]
removing the ux label - this is on our radar already as it's part of Q3 plans.

**Comment by pocmo** [comment]
GV bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1564195

**Comment by liuche** [comment]
This is a Material Component, but there are a few choices to be made - like color, distance of pull. Just requires a few chosen values.

**Comment by mheubusch** [comment]
Details on available decisions: https://developer.android.com/training/swipe/add-swipe-interface

**Comment by AmyYLee** [comment]
@NotWoods do you need a ux review to move ahead with this? Please let me know

**Comment by NotWoods** [comment]
We can move ahead without UX but it would be nice to have suggestions for the color and distance. Our main blocker is GeckoView right now.

**Comment by liuche** [comment]
QA would also like more time to test this, so we should put this behind a Feature Flag.

**Comment by AmyYLee** [comment]
> @NotWoods do you need a ux review to move ahead with this? Please let me know



> We can move ahead without UX but it would be nice to have suggestions for the color and distance. Our main blocker is GeckoView right now.

We should keep the distance of pull unmodified from a stock android component. The only modification would be the colour of the spinner icon.

Default spinner icon colour: #312A65
Private Browsing Mode spinner icon colour: #592ACB 

Let me know if you need anything else

**Comment by liuche** [comment]
GV bug has been fixed, but won't be in GV70, so we'll have to wait until GV71 which merges 10/21.

**Comment by liuche** [comment]
We're now running GV Nightly, so this is ready to be worked on, but needs to be behind a Nightly flag.

**Comment by lime124** [comment]
yay! look forward to it landing in nightly!

@NotWoods can you please let @AmyYLee  review quickly before landing? then we can make any adjustments before it lands. i appreciate it - thanks!

**Comment by jonalmeida** [comment]
@NotWoods removing from our sprint planning board since we're not sure if it's actively being worked on. It's also a user story that doesn't have defined work in it.

**Comment by liuche** [comment]
@NotWoods is there some conclusion in where the responsibilities for where pull-to-refresh lives between GV and AC/Fenix, and also whether that API is Gecko-ish or Android-ish (e.g. params that match Android PTR APIs)?

Let me know if this is blocked on something and we need to make some kind of decision, or what the options are moving forward are (I couldn't tell from the comments in https://github.com/mozilla-mobile/android-components/issues/3182 )

**Comment by NotWoods** [comment]
I've been having a lot of trouble working with the APIs that Gecko provides.

The built in pull to refresh component just checks if we're at the top of the screen, so I've been subclassing it to work around that. It catches touch events early but the order the events are received don't match with what the GeckoView API wants.

I'm also worried that pages where the scrollbar isn't on the body element, such as a large element fixed to the screen size but scrollable, won't work properly since we can only check the body scroll position.

I'm still trying to work on this but I've been focusing on the PWA work so this has been in my backlog.

**Comment by liuche** [comment]
Popping this onto @fluffyemily's radar - in the meantime, I'll also add it to the GV board.

@NotWoods I see a [PTR](https://github.com/mozilla-mobile/android-components/issues/3182) bug on the PWA board, so let me know how important it is to resurface this to get a more helpful API from Gecko.

**Comment by NotWoods** [comment]
PTR for PWA isn't a blocker since we have the refresh button in the site controls notification.

**Comment by snorp** [comment]
> I've been having a lot of trouble working with the APIs that Gecko provides.
> 
> The built in pull to refresh component just checks if we're at the top of the screen, so I've been subclassing it to work around that.

What is the "pull to refresh component"? Are we talking about `androidx.swiperefreshlayout.widget.SwipeRefreshLayout`?

> It catches touch events early but the order the events are received don't match with what the GeckoView API wants.

I'm not sure what you mean by this. Which GeckoView API specifically? If we're talking about `PanZoomController.onTouchEvent()` the order is assumed to be chronological. How are the events getting into the wrong order?

> 
> I'm also worried that pages where the scrollbar isn't on the body element, such as a large element fixed to the screen size but scrollable, won't work properly since we can only check the body scroll position.

Yeah, that's not something we're able to handle right now. I talked to the APZ folks about it a while back and it sounded like it would be a pretty heavy lift.

> 
> I'm still trying to work on this but I've been focusing on the PWA work so this has been in my backlog.

**Comment by Mugurell** [comment]
I was looking into this today and saw that Tiger (and others) did an awesome job and we already have a functional pull down to refresh feature in the app.

But there is still a rather important issue remaining: pages that are already scrolled to the top but contain scrollable content.
Because of the current implementation our `SwipeRefreshLayout` races async with GeckoView to handle the scroll.
On a scroll down we'll trigger the pull down to refresh because the page is already at top.
On a scroll down GV will scroll the scrollable content.
Gecko is right, we are wrong.
This is quite problematic on Google Maps for example.

The current implementation based on a `SwipeRefreshLayout` means a callback will be fired for every scroll event and [we'll just check if the page is already at top](https://github.com/mozilla-mobile/android-components/blob/ce5af2e9ed086440496a0fd3e8e9a1cebf39d96c/components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineView.kt#L153) based on which the refresh will be triggered or not.
We could add more to this check (not clear what atm) but this is already quite expensive and could probably still fail in some cases.

So I though about going the other way around: 
instead of the app having it's own swipe listener that races with Gecko's (and it's clear that it's own would be wrong in some cases while Gecko's wouldn't because we only know about the Gecko**View** state while Gecko's know about website's) why don't we use Gecko's?
In Fennec we already use the [GeckoSession#updateOverscrollVelocity](https://dxr.mozilla.org/mozilla-central/rev/9b338268ce36c3387eb27c15f082e8c8f421112c/mobile/android/geckoview/src/main/java/org/mozilla/geckoview/GeckoSession.java#270) and [GeckoSession#updateOverscrollOffset](https://dxr.mozilla.org/mozilla-central/rev/9b338268ce36c3387eb27c15f082e8c8f421112c/mobile/android/geckoview/src/main/java/org/mozilla/geckoview/GeckoSession.java#275) to calculate and display the overscroll effect (another requested Fenix feature). 
I think we can do the same in Fenix (with some help from a-c and roll our own pull down to refresh and overscroll effect functionality.
This would also work for the above Google Maps example and should also be cheaper in terms of a performance hit.

Curious about @snorp's opinion from a GV perspective.

**Comment by snorp** [comment]
You can detect that Gecko is handing the touch events to web content (as in the case with Google Maps) by handling the events with [`GeckoView.onTouchEventForResult()`](https://mozilla.github.io/geckoview/javadoc/mozilla-central/org/mozilla/geckoview/GeckoView.html#onTouchEventForResult-android.view.MotionEvent-) and observing the return value. `INPUT_RESULT_HANDLED_CONTENT` indicates that web content handled the event and we probably don't want to use it for pull-to-refresh.

Using the overscroll stuff for this is an interesting idea. I'm not sure why the overscroll methods aren't being called currently. It needs some investigation.

**Comment by Mugurell** [comment]
Awesome!
Thanks James!

Sounds like we should have everything we need for pull down to refresh!

**Comment by lindongbin** [comment]
is there an option to disable "pull down refresh"?

**Comment by lindongbin** [comment]
> 
> 
> Please consider having this gesture be optional. I very rarely have a need to refresh a page and find it easy to accidentally trigger refreshes in other apps with pull-to-refresh when scrolling.

the same

**Comment by stefan-niedermann** [comment]
@lindongbin please use the up-/downvote features at [this comment](https://github.com/mozilla-mobile/fenix/issues/3262#issuecomment-500211623) which already asked this question to avoid flooding this thread.

**Comment by spinda** [comment]
The issue described in https://github.com/mozilla-mobile/fenix/issues/3262#issuecomment-603429247 is present on https://5e.tools/spells.html and other sites. Tap "Filter" and scroll through the modal that appears, then try to scroll it back the other way. Pages like this are now unusable on Fenix Nightly, and there's no way to disable. Is there a reason that new Fenix features aren't being assigned `about:config` flags like traditional Firefox ones?

**Comment by madb1lly** [comment]
Hi all,

Pull to refresh is implemented and supposed to be working? Not on my Nokia 8 on Android 9 it isn't. My Fenix build is Nightly 200403 06:01 (Build #2015732339) 38.0.20200402130113, ea47f9c33.

Cheers 🙂

**Comment by cadeyrn** [comment]
> Pull to refresh is implemented and supposed to be working? 

Yes but it's still in development and behind a Nightly flag. It works on many websites but it's not yet working on every website and does not work correctly on some other websites, see the [open issues](https://github.com/mozilla-mobile/fenix/issues?q=is%3Aissue+is%3Aopen+pull+to+refresh).

**Comment by madb1lly** [comment]
Hi @cadeyrn, thanks for the info. After update yesterday it seems to be enabled on my build.
Cheer 🙂

**Comment by abodea** [comment]
Verified as fixed on Nightly 4/7 with Google Pixel 4 XL (10).
Note that the `Pull to Refresh a Web Page` was implemented.

**Comment by Mugurell** [comment]
We'll unfortunately also disable pull to refresh based on #9766 pending changes.
Hope to enable it again in the near future.

**Comment by masayuki-nakano** [comment]
I wonder, do users really need this feature?

There are indeed 2 types of web pages, one is modified frequently, the other is static or rarely modified. I guess that this UI is designed for the former, but such web apps usually have their own refresh UI like Twitter. For the latter case, I belive that this kind of pages are more usual case, this UI is just annoying especially when connecting the Internet with measured rate line...

**Comment by cadeyrn** [comment]
@masayuki-nakano: Most news websites, online communities or platforms like GitHub does _not_ have such a feature. To be honest Twitter seems to be the exception of the rule. This is at least true for the websites I know. Sure, it's not the most important feature since you don't need this to refresh the page but it's still a very nice feature and something that Chrome offers. It's so nice because you save time. Time is probably the most valuable good in the world. ;-)

**Comment by masayuki-nakano** [comment]
@cadeyrn Thank you for explaining the example. Are there some hints to enable this feature only on some web sites? E.g., cache control header, network access by JS or something. I mean that it'd be nicer if this feature was **disabled** by heuristic algorithm. E.g., reloading treehearder, searchfox accidentally makes me really annoying because of too long loading time.

**Comment by madb1lly** [comment]
Hi @mheubusch, @cadeyrn and others

As a user who has now used this feature for a couple of days, I do not want to pull down to refresh the page. At least not using the current implementation.

The idea that it saves time only works if it cannot be triggered accidentally, but unfortunately I find myself frequently triggering it accidentally whenever I scroll back to the top of a page, e.g. to access a menu, breadcrumbs, shopping categories etc. I then need to wait for the page to reload, so I'm actually losing time with this feature not gaining time.

Please add a toggle into settings to turn this feature on or off.

Thanks 🙂

**Comment by cadeyrn** [comment]
You shouldn't mix up bugs in the current implementation with the final implementation which is not yet available. If you want to disable Pull to Refresh because you're triggering it accidentally due to a bug in the current implementation it's a completely different story than if you want to disable the final Pull to Refresh implementation because you don't like the feature at all.

Please see #9766 for open and known issues with the current implementation. Because of these issues Pull to Refresh was already disabled again for all users.

**Comment by cadeyrn** [comment]
@masayuki-nakano:

> Are there some hints to enable this feature only on some web sites? E.g., cache control header, network access by JS or something. I mean that it'd be nicer if this feature was disabled by heuristic algorithm. E.g., reloading treehearder, searchfox accidentally makes me really annoying because of too long loading time.

If the feature only works on some websites it feels broken because I, as a user, don't know and can't control where it works and where not. I think the real issue that it's too easy to trigger it accidentally. So this should be solved. Also an option to completely disable the feature is fine. But disabling a feature for some websites for reasons that are totally intransparent to the user does not seem like a good idea to me.

**Comment by masayuki-nakano** [comment]
@cadeyrn okay, but is it cancelable from web apps which do not want this feature? There shoild be at least one event whose `preventDefault()` makes Fenix stop starting to observe the gesture. And it should be announced via release note at least. (Anyway under current market share, most web developers would not do something for Fenix, though. I, as a DOM Event module owner, really worry about new web-compat issues...)

**Comment by Mugurell** [comment]
Closing as per https://github.com/mozilla-mobile/fenix/issues/3262#issuecomment-610414753

**Comment by andreicristianpetcu** [comment]
https://github.com/mozilla-mobile/fenix/issues/17275
